Testing: Print test reproduction command on failure

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99444
* #104819
* __->__ #104537

MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
* To keep track of flag settings, this PR introduces a new `TestEnvironment` class that defines global flags by querying related environment variables. Flag and env var names are purposefully kept searchable via full names. Example usages:
```python
TestEnvironment.def_flag("TEST_WITH_TORCHINDUCTOR", env_var="PYTORCH_TEST_WITH_INDUCTOR")
# can track implication relationships to avoid adding unnecessary flags to the repro
TestEnvironment.def_flag(
    "TEST_WITH_TORCHDYNAMO",
    env_var="PYTORCH_TEST_WITH_DYNAMO",
    implied_by_fn=lambda: TEST_WITH_TORCHINDUCTOR or TEST_WITH_AOT_EAGER)
# can use include_in_repro=False to keep the flag from appearing in the repro command
TestEnvironment.def_flag(
    "DISABLE_RUNNING_SCRIPT_CHK", env_var="PYTORCH_DISABLE_RUNNING_SCRIPT_CHK", include_in_repro=False)
# the default default value is False, but this can be changed
TestEnvironment.def_flag(
    "PRINT_REPRO_ON_FAILURE", env_var="PYTORCH_PRINT_REPRO_ON_FAILURE", default=(not IS_FBCODE), include_in_repro=False)
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

jbschlosser committed:
Testing: Print test reproduction command on failure

[ghstack-poisoned]

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1038,6 +1040,34 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
+
+    return " ".join([f"{non_standard_mappings[flag]}=1"
+                     if flag in non_standard_mappings
+                     else f"PYTORCH_{flag}=1" for flag in enabled])
```
Since this is a BE patch, I'm going to ask you to do a little more surgery. Instead of hacking up globals() to detect environment variable settings, can we instead move the entire test environment variable apparatus to a class that can keep track of all the test knobs? If your primary objective is to avoid having to edit use sites of things like TEST_WITH_TORCHINDUCTOR, you can still assign these globals; the idea is to just have a clear class that handles *defining* the variables in one place.

jbschlosser replied:
Cool, I'm on-board with this. I actually started down this route initially but decided to keep it simple / hacky in the initial commit.

jbschlosser replied:
This could easily fall prey to scope creep trying to gather all test configuration into a single class. For now, I introduced a simple `TestEnvironment` class that handles environment variable related boolean flags, exporting flags globally to avoid call-site changes. I could also extend it to:
1. Handle non-boolean environment variable settings
2. Handle all test related flags that aren't related to environment variables

We'd be introducing more complexity and changes for 1 and especially for 2. Neither of those are strictly necessary for the repro command use case, and they'd have questionable payoff. wdyt?

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1732,6 +1762,19 @@ def skip_exception_type(exc_type):
+        raise
+    except Exception as e:
+        # NB: Hacking the exception args is the cleanest way I've found to append
+        # failure reproduction info without poisoning the stack trace.
```
I suppose the other possibility is to modify the overall test runner. This is how I added the hint for 'keep-going':

```
        if (
            not options.continue_through_error
            and not RERUN_DISABLED_TESTS
            and len(failure_messages) != 0
        ):  
            raise RuntimeError(
                "\n".join(failure_messages)
                + "\n\nTip: You can keep running tests even on failure by "
                "passing --keep-going to run_test.py.\n"
                "If running on CI, add the 'keep-going' label to "
                "your PR and rerun your jobs."
            )
```

This is done in run_test.py, so it will get printed even if the testing process crashes.

jbschlosser replied:
I'd like to be consistent with that, but I don't think I have access to the right stuff at the `run_test.py` level. AFAICT the test runner isn't aware of individual test case names.

Also it's kinda nice to get the repro output when running outside of a `run_test.py` context as well.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -109,6 +109,8 @@
 OVERRIDE_FLAKY_SIGNAL = os.getenv('PYTORCH_OVERRIDE_FLAKY_SIGNAL') == '1'
 DISABLE_RUNNING_SCRIPT_CHK = os.getenv('PYTORCH_DISABLE_RUNNING_SCRIPT_CHK') == '1'
+# NB: enabled by default!
+PRINT_REPRO_ON_FAILURE = os.getenv('PYTORCH_PRINT_REPRO_ON_FAILURE', '1') == '1'
```
One reason to disable this is if you're running in fbcode (though you're not testing this right now)

jbschlosser replied:
Disabling for fbcode is a good point. When I added the toggle, I was thinking about someone wanting to reduce spam when running locally OR avoiding obscuring test signal if my logic is horribly broken for some unforeseen case.

re fbcode: flagging a follow-up to disable the env var there

ezyang replied:
I guarantee that no one is going to know that they can suppress the output with this envvar unless you tell them about it in the messages.

janeyx99 replied:
There should be an IS_CI flag that would make your option 2 easy, @jbschlosser

jbschlosser replied:
Added directions to suppress the output to the message, but I can also scrap this if you think it's overkill.

jbschlosser replied:
> There should be an IS_CI flag that would make your option 2 easy, @jbschlosser

Awesome, thanks! I ended up defaulting the flag to `not IS_FBCODE`.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2153,6 +2196,21 @@ def __init__(self, method_name='runTest'):
+                        lambda repro_str=repro_str: print_repro_on_failure(repro_str=repro_str))
+                except Exception:
+                    # Don't fail entirely if we can't get the test filename
+                    pass
```
worth logging the exception here, `log.exception("could not print repro string")`

ezyang reviewed (approved):
I don't want to be too picky about what eventually goes in, since it is better to have something rather than nothing, but I think my comments are definitely worth looking into.

jbschlosser committed:
Update on "Testing: Print test reproduction command on failure"


MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

[ghstack-poisoned]

jbschlosser committed:
Update on "Testing: Print test reproduction command on failure"


MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
* To keep track of flag settings, this PR introduces a new `TestEnvironment` class that defines global flags by querying related environment variables. Flag and env var names are purposefully kept searchable via full names. Example usages:
```python
TestEnvironment.def_flag("TEST_WITH_TORCHINDUCTOR", env_var="PYTORCH_TEST_WITH_INDUCTOR")
# can track implication relationships to avoid adding unnecessary flags to the repro
TestEnvironment.def_flag(
    "TEST_WITH_TORCHDYNAMO",
    env_var="PYTORCH_TEST_WITH_DYNAMO",
    implied_by_fn=lambda: TEST_WITH_TORCHINDUCTOR or TEST_WITH_AOT_EAGER)
# can use include_in_repro=False to keep the flag from appearing in the repro command
TestEnvironment.def_flag(
    "DISABLE_RUNNING_SCRIPT_CHK", env_var="PYTORCH_DISABLE_RUNNING_SCRIPT_CHK", include_in_repro=False)
# the default default value is False, but this can be changed
TestEnvironment.def_flag(
    "PRINT_REPRO_ON_FAILURE", env_var="PYTORCH_PRINT_REPRO_ON_FAILURE", default=(not IS_FBCODE), include_in_repro=False)
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

[ghstack-poisoned]

jbschlosser committed:
Update on "Testing: Print test reproduction command on failure"


MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
* To keep track of flag settings, this PR introduces a new `TestEnvironment` class that defines global flags by querying related environment variables. Flag and env var names are purposefully kept searchable via full names. Example usages:
```python
TestEnvironment.def_flag("TEST_WITH_TORCHINDUCTOR", env_var="PYTORCH_TEST_WITH_INDUCTOR")
# can track implication relationships to avoid adding unnecessary flags to the repro
TestEnvironment.def_flag(
    "TEST_WITH_TORCHDYNAMO",
    env_var="PYTORCH_TEST_WITH_DYNAMO",
    implied_by_fn=lambda: TEST_WITH_TORCHINDUCTOR or TEST_WITH_AOT_EAGER)
# can use include_in_repro=False to keep the flag from appearing in the repro command
TestEnvironment.def_flag(
    "DISABLE_RUNNING_SCRIPT_CHK", env_var="PYTORCH_DISABLE_RUNNING_SCRIPT_CHK", include_in_repro=False)
# the default default value is False, but this can be changed
TestEnvironment.def_flag(
    "PRINT_REPRO_ON_FAILURE", env_var="PYTORCH_PRINT_REPRO_ON_FAILURE", default=(not IS_FBCODE), include_in_repro=False)
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

[ghstack-poisoned]

jbschlosser committed:
Update on "Testing: Print test reproduction command on failure"


MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
* To keep track of flag settings, this PR introduces a new `TestEnvironment` class that defines global flags by querying related environment variables. Flag and env var names are purposefully kept searchable via full names. Example usages:
```python
TestEnvironment.def_flag("TEST_WITH_TORCHINDUCTOR", env_var="PYTORCH_TEST_WITH_INDUCTOR")
# can track implication relationships to avoid adding unnecessary flags to the repro
TestEnvironment.def_flag(
    "TEST_WITH_TORCHDYNAMO",
    env_var="PYTORCH_TEST_WITH_DYNAMO",
    implied_by_fn=lambda: TEST_WITH_TORCHINDUCTOR or TEST_WITH_AOT_EAGER)
# can use include_in_repro=False to keep the flag from appearing in the repro command
TestEnvironment.def_flag(
    "DISABLE_RUNNING_SCRIPT_CHK", env_var="PYTORCH_DISABLE_RUNNING_SCRIPT_CHK", include_in_repro=False)
# the default default value is False, but this can be changed
TestEnvironment.def_flag(
    "PRINT_REPRO_ON_FAILURE", env_var="PYTORCH_PRINT_REPRO_ON_FAILURE", default=(not IS_FBCODE), include_in_repro=False)
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

[ghstack-poisoned]

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -94,21 +94,94 @@
+            TestEnvironment.repro_env_vars[name] = env_var_val
+
+        # export flag globally for convenience
+        globals()[name] = enabled
```
Maybe assert name isn't already set?

jbschlosser committed:
Update on "Testing: Print test reproduction command on failure"


MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
* To keep track of flag settings, this PR introduces a new `TestEnvironment` class that defines global flags by querying related environment variables. Flag and env var names are purposefully kept searchable via full names. Example usages:
```python
TestEnvironment.def_flag("TEST_WITH_TORCHINDUCTOR", env_var="PYTORCH_TEST_WITH_INDUCTOR")
# can track implication relationships to avoid adding unnecessary flags to the repro
TestEnvironment.def_flag(
    "TEST_WITH_TORCHDYNAMO",
    env_var="PYTORCH_TEST_WITH_DYNAMO",
    implied_by_fn=lambda: TEST_WITH_TORCHINDUCTOR or TEST_WITH_AOT_EAGER)
# can use include_in_repro=False to keep the flag from appearing in the repro command
TestEnvironment.def_flag(
    "DISABLE_RUNNING_SCRIPT_CHK", env_var="PYTORCH_DISABLE_RUNNING_SCRIPT_CHK", include_in_repro=False)
# the default default value is False, but this can be changed
TestEnvironment.def_flag(
    "PRINT_REPRO_ON_FAILURE", env_var="PYTORCH_PRINT_REPRO_ON_FAILURE", default=(not IS_FBCODE), include_in_repro=False)
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

[ghstack-poisoned]

janeyx99 reviewed (approved):
woo!!

huydhn reviewed (approved):
LGTM!  The lint failure https://github.com/pytorch/pytorch/actions/runs/5490706756/jobs/10006471108 is a legit annoyance though as it uses nightly build to run.  IMO, an easy, but rather slow, work around would be to land the change in common_utils first, then land the change in run_test after

jbschlosser committed:
Update on "Testing: Print test reproduction command on failure"


MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
* To keep track of flag settings, this PR introduces a new `TestEnvironment` class that defines global flags by querying related environment variables. Flag and env var names are purposefully kept searchable via full names. Example usages:
```python
TestEnvironment.def_flag("TEST_WITH_TORCHINDUCTOR", env_var="PYTORCH_TEST_WITH_INDUCTOR")
# can track implication relationships to avoid adding unnecessary flags to the repro
TestEnvironment.def_flag(
    "TEST_WITH_TORCHDYNAMO",
    env_var="PYTORCH_TEST_WITH_DYNAMO",
    implied_by_fn=lambda: TEST_WITH_TORCHINDUCTOR or TEST_WITH_AOT_EAGER)
# can use include_in_repro=False to keep the flag from appearing in the repro command
TestEnvironment.def_flag(
    "DISABLE_RUNNING_SCRIPT_CHK", env_var="PYTORCH_DISABLE_RUNNING_SCRIPT_CHK", include_in_repro=False)
# the default default value is False, but this can be changed
TestEnvironment.def_flag(
    "PRINT_REPRO_ON_FAILURE", env_var="PYTORCH_PRINT_REPRO_ON_FAILURE", default=(not IS_FBCODE), include_in_repro=False)
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

[ghstack-poisoned]

jbschlosser commented:
@huydhn do you have any insight on the other test failures? Doesn't seem related to my stuff (can repro locally without my changes). am I okay to ignore these?

jbschlosser committed:
Update on "Testing: Print test reproduction command on failure"


MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
* To keep track of flag settings, this PR introduces a new `TestEnvironment` class that defines global flags by querying related environment variables. Flag and env var names are purposefully kept searchable via full names. Example usages:
```python
TestEnvironment.def_flag("TEST_WITH_TORCHINDUCTOR", env_var="PYTORCH_TEST_WITH_INDUCTOR")
# can track implication relationships to avoid adding unnecessary flags to the repro
TestEnvironment.def_flag(
    "TEST_WITH_TORCHDYNAMO",
    env_var="PYTORCH_TEST_WITH_DYNAMO",
    implied_by_fn=lambda: TEST_WITH_TORCHINDUCTOR or TEST_WITH_AOT_EAGER)
# can use include_in_repro=False to keep the flag from appearing in the repro command
TestEnvironment.def_flag(
    "DISABLE_RUNNING_SCRIPT_CHK", env_var="PYTORCH_DISABLE_RUNNING_SCRIPT_CHK", include_in_repro=False)
# the default default value is False, but this can be changed
TestEnvironment.def_flag(
    "PRINT_REPRO_ON_FAILURE", env_var="PYTORCH_PRINT_REPRO_ON_FAILURE", default=(not IS_FBCODE), include_in_repro=False)
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

[ghstack-poisoned]

huydhn commented:
> @huydhn do you have any insight on the other test failures? Doesn't seem related to my stuff. I can repro locally without my changes, but then I don't see those failures in the HUD. what should I do?

From what I see, they are all disabled tests and shouldn't be run at all in the PR.  Looking at an example failure, I see it was run as:

```
Executing ['/opt/conda/envs/py_3.9/bin/python', '-bb', 'test_type_hints.py', '--shard-id=0', '--num-shards=1', '-v', '-vv', '-rfEX', '-p', 'no:xdist', '--use-pytest', '-x', '--reruns=2']
```

while the correct command should be:

```
Executing ['/opt/conda/envs/py_3.9/bin/python', '-bb', 'test_type_hints.py', '--shard-id=0', '--num-shards=1', '-v', '-vv', '-rfEX', '-p', 'no:xdist', '--use-pytest', '--sc=test_type_hints_0', '-x', '--reruns=2', '--import-slow-tests', '--import-disabled-tests']
```

So there are several missing parameters:

* `--sc=test_type_hints_0`
* `--import-slow-tests`
* and `--import-disabled-tests`

They are all controlled via env variables I think, so let's double check these parts.

jbschlosser committed:
Update on "Testing: Print test reproduction command on failure"


MS2 of the Reproducible Testing BE initiative. For context, this is the ask:

```
Another thing that would be really great as we start to have more dependent 
systems or types of tests (functorch, dynamo, crossref) would be to have a 
minimally reproducible version of the test (something at the end of the HUD 
comment like: "Run python test/test_file.py -k test_name" but also if you need 
flags, like crossref it would be like "Run <flag to run crossref> python test/..." ). I'll 
often go through the test infra to find the flags that I need to pass when 
something only breaks crossref/dynamo tests.
```

Implementation details:
* Adds a new flag `PRINT_REPRO_ON_FAILURE` that is settable through the environment variable `PYTORCH_PRINT_REPRO_ON_FAILURE=1`
    * **Default is ON but I can be persuaded otherwise**
* When the flag is enabled, our base `TestCase` will wrap the test method in a context manager that catches any non-skip exceptions and appends a repro string to the exception message. The repro includes setting of necessary test flags through env vars. Example:

```
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_WITH_CROSSREF=1 python test/test_ops.py -k test_foo_add_cuda_float32

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```
* To keep track of flag settings, this PR introduces a new `TestEnvironment` class that defines global flags by querying related environment variables. Flag and env var names are purposefully kept searchable via full names. Example usages:
```python
TestEnvironment.def_flag("TEST_WITH_TORCHINDUCTOR", env_var="PYTORCH_TEST_WITH_INDUCTOR")
# can track implication relationships to avoid adding unnecessary flags to the repro
TestEnvironment.def_flag(
    "TEST_WITH_TORCHDYNAMO",
    env_var="PYTORCH_TEST_WITH_DYNAMO",
    implied_by_fn=lambda: TEST_WITH_TORCHINDUCTOR or TEST_WITH_AOT_EAGER)
# can use include_in_repro=False to keep the flag from appearing in the repro command
TestEnvironment.def_flag(
    "DISABLE_RUNNING_SCRIPT_CHK", env_var="PYTORCH_DISABLE_RUNNING_SCRIPT_CHK", include_in_repro=False)
# the default default value is False, but this can be changed
TestEnvironment.def_flag(
    "PRINT_REPRO_ON_FAILURE", env_var="PYTORCH_PRINT_REPRO_ON_FAILURE", default=(not IS_FBCODE), include_in_repro=False)
```
* AFAICT it is only feasible to achieve this from within the test framework rather than at the CI level. This is because CI / `run_test.py` are unaware of individual test cases. Implementing it in our base `TestCase` class has the broadest area of effect, as it's not isolated to e.g. OpInfo tests.
* I couldn't find an easy way to test the logic via `test_testing.py`, as the logic for extracting the test filename doesn't work for generated test classes. I'm open to ideas on testing this, however.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot merge
